### PR TITLE
8297724: Loop strip mining prevents some empty loops from being eliminated

### PR DESCRIPTION
**Pull request by rwestrel**
When an empty loop is found, it's removed and as a consequence the
outer strip mine loop and the safepoint that it contains are also
removed. A counted loop is empty if it has the minimum number of nodes
that a well formed counted loop contains. In some cases, the loop has
extra nodes and the safepoint in the outer loop is the only node that
keeps those extra nodes alive. If the safepoint was to be removed,
then the counted loop would have the minimum number of nodes and be
considered empty. But the safepoint can't be removed until the loop is
considered empty which only happens if it has the minimum of nodes. As
a result, these loops are not removed. Note that now that the loop
strip mining loop nest is constructed even if UseCountedLoopSafepoints
is false, there's a regression where some loops used to be removed as
empty before but not anymore.

The fix I propose is to extend IdealLoopTree::do_remove_empty_loop()
so it handles those cases. If it encounters a loop with no flow
control in the loop body but a number of nodes greater than the
minimum number of nodes, it starts from the extra nodes in the loop
body and follows uses until it finds a side effect, ignoring the
safepoint of the outer loop. If it finds none, then the extra nodes
can be removed and the loop is empty. This also works if the extra
nodes are kept alive by the safepoints of 2 different counted loops
and one can only be proven empty if the other one is as well (and the
other one proven empty if the first one is) and should work even if
there are more than 2 nodes involved..

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297724](https://bugs.openjdk.org/browse/JDK-8297724): Loop strip mining prevents some empty loops from being eliminated

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [76fd3598](https://git.openjdk.org/jdk/pull/11699/files/76fd35988bb29233cf810c74f5340d6f9f41a447)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11699/head:pull/11699` \
`$ git checkout pull/11699`

Update a local copy of the PR: \
`$ git checkout pull/11699` \
`$ git pull https://git.openjdk.org/jdk pull/11699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11699`

View PR using the GUI difftool: \
`$ git pr show -t 11699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11699.diff">https://git.openjdk.org/jdk/pull/11699.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good to me. Thank you for fixing it.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -3587,19 +3587,27 @@ void IdealLoopTree::remove_main_post_loops(CountedLoopNode *cl, PhaseIdealLoop *
...
     return false;   // Dead loop
   }
+  if (!empty_loop_candidate(phase)) {
+    return false;
+  }
   CountedLoopNode *cl = _head->as_CountedLoop();
-  if (!cl->is_valid_counted_loop(T_INT)) {
-    return false;   // Malformed loop
+#ifdef ASSERT
+  // Call collect_loop_core_nodes to exercise the assert that check that it finds the right number of nodes
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopTransform.cpp:3598]:
```suggestion
  // Call collect_loop_core_nodes to exercise the assert that checks that it finds the right number of nodes
```

File: test/hotspot/jtreg/compiler/c2/irTests/TestLSMMissedEmptyLoop.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+
+/*
+ * @test
+ * @library /test/lib /
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/TestLSMMissedEmptyLoop.java:30]:
```suggestion
 * @test
 * @bug 8297724
 * @library /test/lib /
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I'm running some testing and will report back once it passed.

**Review by TobiHartmann — APPROVED** [review]
All tests passed.

**Comment by rwestrel** [comment]
@vnkozlov @TobiHartmann thanks for the reviews (and testing).

**Pull request closed without merging**